File: src/main/java/org/stt/time/DurationRounder.java
package org.stt.time;


import javax.inject.Singleton;
import java.time.Duration;
import java.util.Objects;

@Singleton
public class DurationRounder {

	private long intervalMillis;
	private long tieBreakMillis;

	public void setInterval(Duration interval) {
        Objects.requireNonNull(interval);
        intervalMillis = interval.toMillis();
        tieBreakMillis = intervalMillis / 2;
	}

	public Duration roundDuration(Duration duration) {
        Objects.requireNonNull(duration);

        long millisToRound = duration.toMillis();
        long segments = millisToRound / intervalMillis;
		long result = segments * intervalMillis;
		long delta = millisToRound - result;
		if (delta == tieBreakMillis && segments % 2 == 1 || delta > tieBreakMillis) {
			result += intervalMillis;
		}
        return Duration.ofMillis(result);
    }

}

File: src/main/kotlin/org/stt/time/DurationRounder.kt
package org.stt.time


import java.time.Duration
import javax.inject.Singleton

@Singleton
class DurationRounder {

    private var intervalMillis: Long = 1
    private var tieBreakMillis: Long = 1

    fun setInterval(interval: Duration) {
        intervalMillis = interval.toMillis()
        tieBreakMillis = intervalMillis / 2
    }

    fun roundDuration(duration: Duration): Duration {
        val millisToRound = duration.toMillis()
        val segments = millisToRound / intervalMillis
        var result = segments * intervalMillis
        val delta = millisToRound - result
        if (delta == tieBreakMillis && segments % 2 == 1L || delta > tieBreakMillis) {
            result += intervalMillis
        }
        return Duration.ofMillis(result)
    }

}
